src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java

```diff
@@ -0,0 +1,215 @@
+import co.cask.cdap.api.dataset.lib.PartitionedFileSet;
+import co.cask.cdap.api.dataset.lib.PartitionedFileSetArguments;
+import co.cask.cdap.api.dataset.lib.PartitionedFileSetProperties;
+import co.cask.cdap.api.dataset.lib.Partitioning;
+import co.cask.cdap.api.mapreduce.MapReduceTaskContext;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.hydrator.common.HiveSchemaConverter;
+import co.cask.hydrator.plugin.common.FileSetUtil;
+import co.cask.hydrator.plugin.common.StructuredToAvroTransformer;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.hadoop.io.NullWritable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSink} to write Avro records to a {@link PartitionedFileSet}.
+ */
+@Plugin(type = "batchsink")
+@Name("DynamicPFSAvro")
+@Description("Sink for a PartitionedFileSet that writes data in Avro format and uses a dynamic partition key.")
+public class DynamicPartitionFileSetAvroSink extends
+  PartitionedFileSetSink<AvroKey<GenericRecord>, NullWritable> {
+  public static final String NAME = "DynamicPFSAvro";
+
+  private static final Logger LOG = LoggerFactory.getLogger(DynamicPartitionFileSetAvroSink.class);
+  private static final String SCHEMA_DESC = "The Avro schema of the record being written to the Sink as a JSON " +
+    "Object.";
+  private static final String FIELD_DESC = "The fields to be used for the partitions as comma separated values.";
+
+  private StructuredToAvroTransformer recordTransformer;
+  private final DynamicPartitionAvroSinkConfig config;
+
+  // A prefix is added to the partition name to avoid conflicts with hive table column names
+  private static final String DPFS_PREFIX = "partition";
+
+  public DynamicPartitionFileSetAvroSink(DynamicPartitionAvroSinkConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) {
+    Map<String, String> sinkArgs = getAdditionalPFSArguments();
+    PartitionedFileSetArguments.setDynamicPartitioner(sinkArgs, DynamicPartitionFileSetAvroSink.FieldValueDynamicPartitioner.class);
+    context.addOutput(Output.ofDataset(config.name, sinkArgs));
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    String pfsName = config.name;
+    String basePath = config.basePath == null ? pfsName : config.basePath;
+
+    // parse to make sure it's valid
+    new org.apache.avro.Schema.Parser().parse(config.schema);
```

Comment:
Can you move all the validation logic to a separate validate function in the config class to match many of the other plugins? Also, what happens if the list of partition fields is macro enabled?